<a href="https://colab.research.google.com/github/ckakalou/Komenti_HandsOn/blob/main/Komenti_and_Friends_Tutorial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Introduction

This is the Komenti and Friends tutorial! It will first be given on the 16th of June during the HealTAC 2021 tutorial afternoon. However, it will subsequently be posted online publicly. 

The goal of this tutorial is to introduce you to a set of tools we have developed, that you can use for the semantic investigation and analysis of (particularly) text data. Everything should run inside the notebook. To speed things up, some of the data files - particularly the text annotation files - are pre-created and downloaded, rather than generated in here, but the commands you can use to generate the files are also given.

Many thanks to CogStack and friends for the text data we will be using, which was used throughout their previous tutorial and can be found in their [github](https://github.com/CogStack/MedCAT/) repository.


# Part One: Text Annotations and Phenotype Profiles

In this section, we will load some text, examine it, annotate it. We will then discuss how these annotations can serve as phenotype profiles. These phenotype profiles will constitute the data that we will use in the subsequent chapters!

First we will download and set up all of the software/files needed... This should only take a minute or so:

In [ ]:
#Download Komenti
!rm -rf komenti
!rm komenti-0.2.0-SNAPSHOT.zip
!wget https://github.com/reality/komenti/releases/download/0.2.0-SNAPSHOT-4/komenti-0.2.0-SNAPSHOT.zip
!unzip komenti-0.2.0-SNAPSHOT.zip
!mv komenti-0.2.0-SNAPSHOT komenti

#Download Klarigi
!rm -rf klarigi
!rm klarigi-0.0.8.zip
!wget https://github.com/reality/klarigi/releases/download/0.0.8/klarigi-0.0.8.zip
!unzip klarigi-0.0.8.zip
!mv klarigi-0.0.8 klarigi

#Download the note events, and the vocab, and the annotations, and the HPO
!rm noteevents.csv
!wget https://raw.githubusercontent.com/CogStack/MedCAT/master/tutorial/data/noteevents.csv
!rm annotations.txt
!wget http://lokero.xyz/annotations.txt
!wget https://raw.githubusercontent.com/reality/synonym_expansion_validation/master/hpo/unexpanded_all.txt
!mv unexpanded_all.txt hpo_vocabulary.txt
!wget http://purl.obolibrary.org/obo/hp.owl

!rm notecat_sim.txt
!wget http://lokero.xyz/notecat_sim.txt

!rm -rf sample_data
!mkdir texts

rm: cannot remove 'komenti-0.2.0-SNAPSHOT.zip': No such file or directory
--2021-06-16 14:18:28--  https://github.com/reality/komenti/releases/download/0.2.0-SNAPSHOT-4/komenti-0.2.0-SNAPSHOT.zip
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://github-releases.githubusercontent.com/225223734/d460ae80-a82e-11eb-910e-5f96167c2c3f?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIWNJYAX4CSVEH53A%2F20210616%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20210616T141828Z&X-Amz-Expires=300&X-Amz-Signature=19ed63fed3173a58a070ea88e68ffc2b77d27ab2ebbe366c257ebc2eadfc9eee&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=225223734&response-content-disposition=attachment%3B%20filename%3Dkomenti-0.2.0-SNAPSHOT.zip&response-content-type=application%2Foctet-stream [following]
--2021-06-16 14:18:28--  https://github-releases.githubusercontent.com/225223734/d4

We're going to be using these notevents data:

In [ ]:
!head noteevents.csv

,subject_id,chartdate,category,text
0,0,01/01/2086, Urology,"CHIEF COMPLAINT: , Blood in urine.,HISTORY OF PRESENT ILLNESS:  ,This is a 78-year-old male who has prostate cancer with metastatic disease to his bladder and in several locations throughout the skeletal system including the spine and shoulder.  The patient has had problems with hematuria in the past, but the patient noted that this episode began yesterday, and today he has been passing principally blood with very little urine.  The patient states that there is no change in his chronic lower back pain and denies any incontinence of urine or stool.  The patient has not had any fever.  There is no abdominal pain and the patient is still able to pass urine.  The patient has not had any melena or hematochezia.  There is no nausea or vomiting.  The patient has already completed chemotherapy and is beyond treatment for his cancer at this time.  The patient is receiving radiation therapy, but it is targeted to the bones and intended

Let's start by loading each of the notes into a dictionary, storing the associated text and structural data together:

In [ ]:
# Import necessary libraries for reading CSV files
import csv
import string

records_by_item = dict()
with open('noteevents.csv') as data:
  reader = csv.DictReader(data)
  for row in reader:
    # Remove trailing space from the category
    row['category'] = row['category'].strip()

    # row[''] refers to the canonical note ID (which is simply an incremental integer)
    records_by_item[row['']] = row
    
print('Done!')

Done!


Now we have a dictionary containing every note event. Let's take a look at one:

In [ ]:
records_by_item['989']

OrderedDict([('', '989'),
             ('subject_id', '137'),
             ('chartdate', '01/01/2062'),
             ('category', 'Consult - History and Phy.'),
             ('text',
              "CHIEF COMPLAINT:, Blood-borne pathogen exposure., ,HISTORY OF PRESENT ILLNESS: ,The patient is a 54-year-old right-handed male who works as a phlebotomist and respiratory therapist at Hospital. The patient states that he was attempting to do a blood gas. He had his finger of the left hand over the pulse and was inserting a needle using the right hand. He did have a protective clothing including use of gloves at the time of the incident. As he advanced the needle, the patient jerked away, this caused him to pull out of the arm and inadvertently pricked the tip of his index finger. The patient was seen and evaluated at the emergency department at the time of incident and had baseline studies drawn, and has been followed by employee health for his injury. The source patient was tested for signs

We could also group the notes by different means, for example by category or subject ID. Most often in our work, we're looking at the totality of a record for a particular patient, or a subject throughout a particular time point, e.g. an admission. Because we're working with an example dataset, and to keep things simple, we're just going to consider every note separately, but keep in mind that in a real environment you would come across problems with bias doing it this way, e.g. that different notes would concern the same people.

So, onto annotation. Komenti requires text files to consume. In this next cell, we will populate the texts/ directory with one file per note.

> *Bonus information: You could actually simply get it to annotate the CSV directly (look into the --per-line argument), but this makes things clearer and easier to understand.*

In [ ]:
# Iterate the text records
for key in records_by_item:
  item = records_by_item[key]
  # Create a new file in the texts/ directory, with the name [note id].txt
  filename = 'texts/' + key + '.txt'
  with open(filename, 'w') as f:
    f.write(item['text'])

print('Done!')

Done!


So, we have our clinical text organised nicely into files, now. The other thing we need to annotate those texts is a vocabulary. A vocabulary tells a text mining program what to recognise in the text. In this tutorial, we're going to be using the entirety of the Human Phenotype Ontology to annotate the text.

The [Human Phenotype Ontology](https://hpo.jax.org/app/) describes phenotypic abnormalities in humans. It is widely appreciated and utilised for its ontological features: axioms, domain vocabulary, and more.

You can use Komenti to build vocabularies from most biomedical ontologies using the *query* command. We're going to use a pre-created vocabulary here, but I've left in the command to create it, for reference: the command tells komenti to retrieve all subclasses of the concept *Phenotypic abnormality* in HPO, and their labels. Then, we're going to have a quick look at the vocabulary:

In [ ]:
# komenti query -q "'Phenotypic abnormality'" --ontology HP --out hpo_vocabulary.txt

!head hpo_vocabulary.txt

tertiary hyperparathyroidism	http://purl.obolibrary.org/obo/HP_0011770	'Phenotypic abnormality'	HP	1
autoimmune hypoparathyroidism	http://purl.obolibrary.org/obo/HP_0011771	'Phenotypic abnormality'	HP	1
broad proximal phalanx of the 3rd finger	http://purl.obolibrary.org/obo/HP_0009450	'Phenotypic abnormality'	HP	1
broad innermost bone of middle finger	http://purl.obolibrary.org/obo/HP_0009450	'Phenotypic abnormality'	HP	1
abnormality of thyroid morphology	http://purl.obolibrary.org/obo/HP_0011772	'Phenotypic abnormality'	HP	1
abnormal shape of thyroid gland	http://purl.obolibrary.org/obo/HP_0011772	'Phenotypic abnormality'	HP	1
uninodular goiter	http://purl.obolibrary.org/obo/HP_0011773	'Phenotypic abnormality'	HP	1
thyroid follicular adenoma	http://purl.obolibrary.org/obo/HP_0011774	'Phenotypic abnormality'	HP	1
thyroid macrofollicular adenoma	http://purl.obolibrary.org/obo/HP_0011775	'Phenotypic abnormality'	HP	1
thyroid microfollicular adenoma	http://purl.obolibrary.org/obo/HP_00117

So, *hpo_vocabulary.txt* contains a vocabulary for all phenotypic abnormalities in HPO. There is one entry per label that can be recognised in text. The columns are:

1. The label, 
2. The Internationalised Resource Identifier (IRI). Looks like a web link. This is also called a term identifier, and it's just a string that uniquely identifies a biomedical concept in an ontology
3. The group this term is in (in this case, and by default, this is what was passed to the --query argument of komenti above)
4. The ontology this term belongs to.
5. Match priority.

So for example, above we can see that a concept describing abnormal thyroid morphology has two entries with the labels "abnormality of thyroid morphology" and "abnormal shape of thyroid gland." When we annotate a text using this vocabulary, either of those labels appearing in text would be linked to the same ontology class for that condition, identified by this unique IRI (Internationalised Resource Identifier): http://purl.obolibrary.org/obo/HP_0011772

> *Bonus information: the komenti query command has a rich set of features for creating text mining vocabularies. You can use it to create vocabularies from subsets of one ontology, several ontologies, from one ontology but using a novel method to derive synonyms from several ontologies, or query for complex class descriptions using description logic queries! You can read more about the query command and see some examples of its use and parameters in the [Komenti documentation](https://github.com/reality/komenti/blob/master/README.md)*



Now, we *could* run the *komenti annotate* command to annotate our texts, identifying instances. However, in this online environment, it is a little slow, and we have a lot to cover in this tutorial, so we're going to use a pre-populated annotation file. I will list the command here for reference:

> *Bonus information: You can also use the --threads argument to use the power of multi-processing in annotation tasks, e.g. --threads 5*

In [ ]:
#!komenti/bin/komenti annotate -t texts/ -l hpo_vocabulary.txt --out annotations.txt --verbose

Now we can take a look at the annotations:

In [ ]:
!head annotations.txt

974.txt	http://purl.obolibrary.org/obo/HP_0012498	nuchal cord	nuchal cord	'Phenotypic abnormality'		5	nuchal cord x1 were tight and reduced.
798.txt	http://purl.obolibrary.org/obo/HP_0003002	breast carcinoma	breast cancer	'Phenotypic abnormality'		2	patient has had diagnosis of right breast cancer in 1984 with subsequent radiation therapy.
798.txt	http://purl.obolibrary.org/obo/HP_0003002	breast carcinoma	breast cancer	'Phenotypic abnormality'		3	the patient's sister was also diagnosed with breast cancer at the age of 59.
387.txt	http://purl.obolibrary.org/obo/HP_0001250	seizures	seizure	'Phenotypic abnormality'	uncertain	1	xyz street,city, state,dear dr. cd:,thank you for seeing mr. xyz, a pleasant 19-year-old male who has seen you in 2005 for suspected seizure activity.
387.txt	http://purl.obolibrary.org/obo/HP_0001250	seizures	seizure	'Phenotypic abnormality'	negated	2	he comes to my office today continuing on dilantin 300 mg daily and has been seizure episode free for the past 2 1/

So we can see here that we have our annotations. There is one line per concept recognised in a text file. The columns are:

1. Filename
2. IRI; term identifier; looks like a web link
3. Concept label
4. Label recognised in the text 
5. Concept group
6. Context disambiguation tags. So, some are negated mentions, or uncertain
7. Sentence id
8. Sentence text

So, we can now use these annotations to build phenotype profiles, with which we will further explore and try to find some interesting insights about throughout the rest of the tutorial. First, let's load the annotations into phenotype profiles in Python:

In [ ]:
phenotype_profiles = dict()
with open('annotations.txt') as data:
  reader = csv.reader(data, delimiter='\t')
  for row in reader:
    # We recapitulate the note ID here by slicing off the '.txt' suffix
    text_id = row[0].split('.')[0]
    if not text_id in phenotype_profiles:
      phenotype_profiles[text_id] = list()
    
    # Here we build a dict that records the salient information for each annotation 
    item = { 'iri': row[1], 'label': row[2], 'tags': row[5].split(',') }
    phenotype_profiles[text_id].append(item)
    
print('done!')

done!


So now we have a dictionary of the concepts associated with each note. This consists a set of phenotypes associated with that note - as such, we call this a phenotype profile! Let's take a look at one:

In [ ]:
phenotype_profiles['1351']

[{'iri': 'http://purl.obolibrary.org/obo/HP_0002090',
  'label': 'pneumonia',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0012764',
  'label': 'orthopnea',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0002094',
  'label': 'dyspnea',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0012735',
  'label': 'cough',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0005110',
  'label': 'atrial fibrillation',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0001903',
  'label': 'anemia',
  'tags': ['negated']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0000952',
  'label': 'jaundice',
  'tags': ['negated']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0030828',
  'label': 'wheezing',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0030830',
  'label': 'crackles',
  'tags': ['']},
 {'iri': 'http://purl.obolibrary.org/obo/HP_0000112',
  'label': 'nephropathy',
  'tags': ['']}]

You can see that we've recorded the IRI or concept identifier, the concept label, and any of the concept disambiguation tags associated with the annotations.

#Part Two: Semantic Characterisation of Groups of Phenotype Profiles

Now that we have some text-derived phenotype profiles, we can start to leverage semantics to evaluate and explore them. To start with, let's look at a bit of a simple initial research question, based on some of the only categorical data we have associated with our texts:

> *Can we identify the phenotypes that uniquely identify and characterise neurology text notes?* 

To do this, we're going to use another one of our programs: [Klarigi](https://github.com/reality/klarigi/). Klarigi uses a novel multi objective optimisation algorithm to find semantic explanations for groupings of any kinds of entities that are associated with ontology terms. 

Our first step on this journey is to format our phenotype profiles into a data file that describes their groupings. Like so:

In [ ]:
filename = 'groups_by_note_category.tsv'
with open(filename, 'w') as f:
  for key in phenotype_profiles:
    # Select the IRIs in this phenotype profile
    iris = [item['iri'] for item in phenotype_profiles[key]]

    # Identify the group for this phenotype profile: in this case, the note category
    group = records_by_item[key]['category']

    # Now we write the entry for this phenotype profile to the output file
    line = "\t".join((key, ";".join(iris), group))
    f.write(line + '\n')

print('Done!')

Done!


Let's take a look at the new *groups_by_note_category.tsv* file:

In [ ]:
!head groups_by_note_category.tsv

974	http://purl.obolibrary.org/obo/HP_0012498	Surgery
798	http://purl.obolibrary.org/obo/HP_0003002;http://purl.obolibrary.org/obo/HP_0003002	Obstetrics / Gynecology
387	http://purl.obolibrary.org/obo/HP_0001250;http://purl.obolibrary.org/obo/HP_0001250	Neurology
578	http://purl.obolibrary.org/obo/HP_0007185;http://purl.obolibrary.org/obo/HP_0007185;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0001259	Emergency Room Reports
222	http://purl.obolibrary.org/obo/HP_0000217;http://purl.obolibrary.org/obo/HP_0002020;http://purl.obolibrary.org/obo/HP_0000737;http://purl.obolibrary.org/obo/HP_0002099;http://purl.obolibrary.org/obo/HP_0004398;http://purl.obolibrary.org/obo/HP_0005263;http://purl.obolibrary.org/obo/HP_0012393;http://purl.obolibrary.org/obo/HP_0500093;http://purl.obolibrary.org/obo/HP_0012393;http://purl.obolibrary.org/obo/HP_0012393	SOAP / Chart / Progress Notes
229	http://purl.obolibrary.org/obo/HP_0000217;http://purl.obolibrary.org/obo/HP_0002020

So, in the first column we have the note IDs, in the second we have a semicolon-delimited list of ontology identifiers, and in the third we have the category. The third column is the important part here: this identifies the group for each profile.

Klarigi uses the ontology, in this case HPO, and the set of instances, in this case phenotype profiles for clinical note events, to identify characteristic phenotypes for groups (in this case note categories) of instances.

So, let's use Klarigi to take a look at what is characteristic of neurology notes (this should take about 2 minutes):

In [ ]:
!klarigi/bin/klarigi --data groups_by_note_category.tsv --ontology hp.owl --group Neurology

Neurology: Scoring completed. Candidates: 999
----------------
Group: Neurology (86 members)
Overall inclusion: 100.0%
Overall exclusion: 99.95%
Explanatory classes:
  IRI: Abnormal axial skeleton morphology (http://purl.obolibrary.org/obo/HP_0009121), Inclusivity: 0.34, Exclusivity: 0.74, IC: 0.46
  IRI: Seizure (http://purl.obolibrary.org/obo/HP_0001250), Inclusivity: 0.31, Exclusivity: 0.94, IC: 0.47
  IRI: Abnormality of higher mental function (http://purl.obolibrary.org/obo/HP_0011446), Inclusivity: 0.3, Exclusivity: 0.92, IC: 0.53
  IRI: Abnormal homeostasis (http://purl.obolibrary.org/obo/HP_0012337), Inclusivity: 0.31, Exclusivity: 0.49, IC: 0.47
  IRI: Behavioral abnormality (http://purl.obolibrary.org/obo/HP_0000708), Inclusivity: 0.41, Exclusivity: 0.71, IC: 0.5
  IRI: Pain (http://purl.obolibrary.org/obo/HP_0012531), Inclusivity: 0.65, Exclusivity: 0.38, IC: 0.52
  IRI: Abnormality of the vertebral column (http://purl.obolibrary.org/obo/HP_0000925), Inclusivity: 0.31, Exclu

So, how do we interpret this?

There are 86 notes in the Neurology category:

- 100% of them contain at least one of the phenotypes or their subclasses in this list
- 99.95% of other note categories are excluded by at lesat one of these HPs
- 31% of Neurology notes discuss seizures, while only 6% of non-neurology notes do this! I guess it makes sense. 
- While pain is discussed in 65% of cases, it's also not discussed in other rates at a rate of 38%. When inclusivity and exclusivity scores add up to around one, this shows that the rate of appearance in this group is around the same as in all (considered together)
- What else is implied by this output?

So, this is kind of interesting, but we can make this a little spicier. What if we want to explore the characteristic phenotypes for text records that discuss particular diseases? Let's take a look at *atrial fibrillation*

We just have to generate our file to input into Klarigi, using our annotations:

In [ ]:
# Here we define the phenotype that we're interested in. We set the label and the term identifier.
phenotype = {'label': 'atrial fibrillation', 'iri': 'http://purl.obolibrary.org/obo/HP_0005110'}
filename = 'individual_phenotype_group.tsv'

with open(filename, 'w') as f:
  for key in phenotype_profiles:
    # Extract the set of ontology terms associated with this phenotype profile
    iris = [item['iri'] for item in phenotype_profiles[key]]

    # We're only interested in phenotype profiles that contain the phenotype we're interested in (defined above)
    if phenotype['iri'] in iris:
      # We remove the phenotype we're interested in from the phenotype profile, so it doesn't dominate the characterisation
      iris = [item for item in iris if item != phenotype['iri']]

      # Now we write this entry to the output file
      line = "\t".join((key, ";".join(iris), phenotype['label']))
      f.write(line + '\n')

print('Done!')

Done!


Now let's check that we have generated the Klarigi input file correctly:

In [ ]:
!head individual_phenotype_group.tsv

1351	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0001903;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	atrial fibrillation
1233	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0001903;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	atrial fibrillation
2059	http://purl.obolibrary.org/obo/HP_0002381;http://purl.obolibrary.org/obo/HP_0002381;http://purl.obolibrary.org/obo/HP_0000010;http://purl.obolibrary.org/obo/HP_0000010;http://purl.obolibrar

Now we can run Klarigi to identify an explanatory set (this should take about 1m30s):

In [ ]:
!klarigi/bin/klarigi --data individual_phenotype_group.tsv --ontology hp.owl

atrial fibrillation: Scoring completed. Candidates: 284
----------------
Group: atrial fibrillation (86 members)
Overall inclusion: 96.51%
Overall exclusion: 100.0%
Explanatory classes:
  IRI: Cough (http://purl.obolibrary.org/obo/HP_0012735), Inclusivity: 0.31, Exclusivity: 1.0, IC: 0.66
  IRI: Abnormal erythroid lineage cell morphology (http://purl.obolibrary.org/obo/HP_0012130), Inclusivity: 0.37, Exclusivity: 1.0, IC: 0.68
  IRI: Jaundice (http://purl.obolibrary.org/obo/HP_0000952), Inclusivity: 0.31, Exclusivity: 1.0, IC: 0.77
  IRI: Dermatological manifestations of systemic disorders (http://purl.obolibrary.org/obo/HP_0001005), Inclusivity: 0.31, Exclusivity: 1.0, IC: 0.69
  IRI: Cholestasis (http://purl.obolibrary.org/obo/HP_0001396), Inclusivity: 0.31, Exclusivity: 1.0, IC: 0.69
  IRI: Stroke (http://purl.obolibrary.org/obo/HP_0001297), Inclusivity: 0.4, Exclusivity: 1.0, IC: 0.73
  IRI: Chills (http://purl.obolibrary.org/obo/HP_0025143), Inclusivity: 0.34, Exclusivity: 1.0, IC

So, what can we tell from this? We should note that the exclusivity score here is always 100%, or 1, because there are no other groups being considered!

 

So, we're able to look into characterisations of phenotype profiles associated with singular phenotypes. Now, let's look back into comparing different phenotypes. What if we want to contrast two similar phenotypes? The clinical relevance of such a task is clear: many diseases, especially in intensive care environments, can easily be conflated or misdiagnosed - especially if they have a similar phenotypic presentation. We can use data-led exploratory investigations to discover the similarities and differences between these diseases. If we can do this really well, we can inform policy, and provide insights that can aid clinical decisions for diagnosis of commonly misdiagnosed diseases. This is one way that retroactive study of text data can help give us impactful insights that could improve care.

Let's look at two diseases that may present similarly: atrial fibrillation and hypertension. First, we need to set up our input file for Klarigi:

In [ ]:
filename = 'two_phen_groups.tsv'
p1 = {'label': 'atrial fibrillation', 'iri': 'http://purl.obolibrary.org/obo/HP_0005110'}
p2 = {'label': 'hypertension', 'iri': 'http://purl.obolibrary.org/obo/HP_0000822'}

with open(filename, 'w') as f:
  # Iterate every phenotype profile
  for key in phenotype_profiles:
    # Build a list of term IDs, or IRIs, associated with the patient phenotype profile
    iris = [item['iri'] for item in phenotype_profiles[key]]
  
    # We're only interested if one of our two phenotypes is associated with this phenotype profile
    if p1['iri'] in iris or p2['iri'] in iris:

      # However, we're not interested in phenotype profiles that express both 
      if not (p1 in iris and p2 in iris):

        # Determine which group this profile is in
        group = p1['label']
        if p2['iri'] in iris:
          group = p2['label']
          
        # Remove the group IRIs from the list of terms associated with the profile (so they don't skew the results)
        iris = [item for item in iris if item != p1['iri'] and item != p2['iri']]

        # Write a line for this phenotype profile to the output file
        line = "\t".join((key, ";".join(iris), group))
        f.write(line + '\n')
        
!head two_phen_groups.tsv

1351	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0001903;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	atrial fibrillation
1233	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0001903;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	atrial fibrillation
1967	http://purl.obolibrary.org/obo/HP_0001541;http://purl.obolibrary.org/obo/HP_0002202;http://purl.obolibrary.org/obo/HP_0000790;http://purl.obolibrary.org/obo/HP_0001635;http://purl.obolibrar

In [ ]:
filename = 'two_phen_groups2.tsv'
p1 = {'label': 'atrial fibrillation', 'iri': 'http://purl.obolibrary.org/obo/HP_0005110'}
p2 = {'label': 'anemia', 'iri': 'http://purl.obolibrary.org/obo/HP_0001903'}

with open(filename, 'w') as f:
  # Iterate every phenotype profile
  for key in phenotype_profiles:
    # Build a list of term IDs, or IRIs, associated with the patient phenotype profile
    iris = [item['iri'] for item in phenotype_profiles[key]]
  
    # We're only interested if one of our two phenotypes is associated with this phenotype profile
    if p1['iri'] in iris or p2['iri'] in iris:

      # However, we're not interested in phenotype profiles that express both 
      if not (p1 in iris and p2 in iris):

        # Determine which group this profile is in
        group = p1['label']
        if p2['iri'] in iris:
          group = p2['label']
          
        # Remove the group IRIs from the list of terms associated with the profile (so they don't skew the results)
        iris = [item for item in iris if item != p1['iri'] and item != p2['iri']]

        # Write a line for this phenotype profile to the output file
        line = "\t".join((key, ";".join(iris), group))
        f.write(line + '\n')
        
!head two_phen_groups2.tsv

1351	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	anemia
1233	http://purl.obolibrary.org/obo/HP_0002090;http://purl.obolibrary.org/obo/HP_0012764;http://purl.obolibrary.org/obo/HP_0002094;http://purl.obolibrary.org/obo/HP_0012735;http://purl.obolibrary.org/obo/HP_0000952;http://purl.obolibrary.org/obo/HP_0030828;http://purl.obolibrary.org/obo/HP_0030830;http://purl.obolibrary.org/obo/HP_0000112	anemia
1350	http://purl.obolibrary.org/obo/HP_0003418;http://purl.obolibrary.org/obo/HP_0002014;http://purl.obolibrary.org/obo/HP_0002013;http://purl.obolibrary.org/obo/HP_0002027;http://purl.obolibrary.org/obo/HP_0000822;http://purl.obolibrary.org/obo/HP_0012531;http://purl.obolibrary.org/obo/HP_0012531;http:

So, we've built our *two_phen_groups.tsv* file with record profiles stratified by whether they have AF or hypertension. Let us now use Klarigi to identify and contrast their characteristic phenotypes (this should take about 2 minutes):

In [ ]:
!klarigi/bin/klarigi --data two_phen_groups.tsv --ontology hp.owl

atrial fibrillation: Scoring completed. Candidates: 725
hypertension: Scoring completed. Candidates: 725
----------------
Group: atrial fibrillation (52 members)
Overall inclusion: 100.0%
Overall exclusion: 99.83%
Explanatory classes:
  IRI: Respiratory tract infection (http://purl.obolibrary.org/obo/HP_0011947), Inclusivity: 0.48, Exclusivity: 0.87, IC: 0.62
  IRI: Orthopnea (http://purl.obolibrary.org/obo/HP_0012764), Inclusivity: 0.48, Exclusivity: 0.95, IC: 0.78
  IRI: Dyspnea (http://purl.obolibrary.org/obo/HP_0002094), Inclusivity: 0.48, Exclusivity: 0.85, IC: 0.64
  IRI: Cough (http://purl.obolibrary.org/obo/HP_0012735), Inclusivity: 0.52, Exclusivity: 0.81, IC: 0.66
  IRI: Jaundice (http://purl.obolibrary.org/obo/HP_0000952), Inclusivity: 0.52, Exclusivity: 0.94, IC: 0.77
  IRI: Dermatological manifestations of systemic disorders (http://purl.obolibrary.org/obo/HP_0001005), Inclusivity: 0.52, Exclusivity: 0.76, IC: 0.69
  IRI: Cholestasis (http://purl.obolibrary.org/obo/HP_0001

In [ ]:
!klarigi/bin/klarigi --data two_phen_groups2.tsv --ontology hp.owl

anemia: Scoring completed. Candidates: 435
atrial fibrillation: Scoring completed. Candidates: 435
----------------
Group: anemia (164 members)
Overall inclusion: 98.78%
Overall exclusion: 98.33%
Explanatory classes:
  IRI: Jaundice (http://purl.obolibrary.org/obo/HP_0000952), Inclusivity: 0.37, Exclusivity: 0.83, IC: 0.77
  IRI: Wheezing (http://purl.obolibrary.org/obo/HP_0030828), Inclusivity: 0.38, Exclusivity: 1.0, IC: 0.74
  IRI: Back pain (http://purl.obolibrary.org/obo/HP_0003418), Inclusivity: 0.35, Exclusivity: 1.0, IC: 0.76
  IRI: Vomiting (http://purl.obolibrary.org/obo/HP_0002013), Inclusivity: 0.53, Exclusivity: 0.83, IC: 0.74
  IRI: Hypertension (http://purl.obolibrary.org/obo/HP_0000822), Inclusivity: 0.52, Exclusivity: 0.58, IC: 0.72
  IRI: Nausea (http://purl.obolibrary.org/obo/HP_0002018), Inclusivity: 0.39, Exclusivity: 0.78, IC: 0.8
  IRI: Renal insufficiency (http://purl.obolibrary.org/obo/HP_0000083), Inclusivity: 0.45, Exclusivity: 0.6, IC: 0.72
----------------


In this way, we can start to gain some real insight into the differences between these diseases, and their phenotypes. Nice! Add some stuff about interpretation here (it's actually genuinely quite interesting)

> Bonus tasks:
> * Try modifying the code to examine another phenotype, or another pair of phenotypes. You can explore the HPO hierarchy to find additional terms to explore [here](https://www.ebi.ac.uk/ols/ontologies/hp/terms?iri=http%3A%2F%2Fpurl.obolibrary.org%2Fobo%2FHP_0000118&viewMode=PreferredRoots&siblings=false)
> * What characterises other note categories apart from Neurology?
> * Can you modify the code that writes the input file for klarigi to exclude negated or uncertain phenotypes? How does this modify the results? 

#Part Three: Semantic Similarity, Ranking, and Classification

Now we're going to see how we can use semantic similarity to perform ranking and classification tasks. Let's start with our previous example: atrial fibrillation vs hypertension. We've just seen, that patients with these phenotypes have different characteristic phenotypes. Can we exploit these characteristic differences between the two sets in a classification task?

We're going to use semantic similarity to predict, based on phenotype profiles, which of the hypertension and atrial fibrillation groups a note is in (or whatever you have set p1 and p2 to above). We can do this, because we have removed the actual mention of hypertension and atrial fibrillation from the phenotype profiles.

First, we're going to use klarigi to calculate semantic similarity between all of our hypertension and atrial fibrillation phenotype profiles (this should take about 2 minutes):


In [ ]:
!klarigi/bin/klarigi --similarity-mode --resnik-ic --data two_phen_groups.tsv --ontology hp.owl --output two_phen_sim.txt
!head two_phen_sim.txt

This command has taken every atrial fibrillation and hypertension phenotype profile, and has compared it to every other phenotype profile using semantic similarity. This gives us a score between 0 and 1 that measures how similar those phenotype profiles are. We call this a 'pairwise similarity matrix.'

What we're now going to try to do, is for each phenotype profile (representing a note), we're going to turn the set of similarity measurements into a prediction of which group the note is in. There are many ways that this can be done, and can be combined with other methods... For now, we will simply take the most similar note, per semantic similarity, and use the group of that note as the prediction:

In [ ]:
topScores = dict()

# This function identifies, given a phenotype profile, which of the two phenotype 
#   groups it belongs to (according to the p1 and p2 variables we defined earlier)
def identify_group(profile):
  iris = [item['iri'] for item in profile]
  group = p1['label']
  if p2['iri'] in iris:
    group = p2['label']
  return group

with open('two_phen_sim.txt') as data:
  reader = csv.reader(data, delimiter='\t')
  for row in reader:
    # Identify which phenotype group each of the phenotype profiles compared by this line are in 
    n1Group = identify_group(phenotype_profiles[row[0]])
    n2Group = identify_group(phenotype_profiles[row[1]])

    item = { 'n1': row[0], 'n2': row[1], 'score': row[2], 'c1': n1Group, 'c2': n2Group}

    # If we don't have an entry for n1, the current phenotype profile being 
    #  considered, create an try for it in the dictionary of top scores
    if not item['n1'] in topScores:
      topScores[item['n1']] = item
    
    # If the score of the current comparison is greater than the existing top score recorded 
    #   for this phenotype profile, replace it with this one
    if item['score'] > topScores[item['n1']]['score']:
      topScores[item['n1']] = item

print('Done!')

Now we can calculate the performance:

In [ ]:
import sklearn.metrics

# Here we're creating two lists, both with one entry for each phenotype profile:
#  truth: which group does the phenotype profile belong to?
#  pred: which group is this phenotype profile predicted to belong to (i.e. the group of its most similar phenotype profile)?
truth = list()
pred = list()

for k in topScores:
  s = topScores[k]
  truth.append(s['c1'])
  pred.append(s['c2'])

# Calculate precision and recall scores
precision = sklearn.metrics.precision_score(truth, pred, average="weighted")

print('Precision: ' + str(precision))


At 0.98 precision, it's pretty good. Obviously we don't have a training/test set here, but we're also not doing any training - in a practical environment, we'd think about reserving a cache of phenotype profiles that new admissions would be compared to for automatic coding, classification, etc.

What if we try to predict something more advanced, like what the category of a note is? We have to calculate similarity again, but we're going to skip that for the purposes of this tutorial, since it takes a little while (especially on this platform). Here is the code to generate it:

In [ ]:
#!klarigi/bin/klarigi --similarity-mode --resnik-ic --data groups_by_note_category.tsv --ontology hp.owl --output notecat_sim.txt
!head notecat_sim.txt

Now let's build our predictions dictionary, by again using the closest partner for each phenotype profile as our predictive label:

In [ ]:
topScores = dict()

with open('notecat_sim.txt') as data:
  reader = csv.reader(data, delimiter='\t')
  for row in reader:
    # In this case, the label we're trying to predict is the note category. So,
    #  to set the 'group' for our two comparisons here, we're going back to 
    #  our dictionary of note events, and extracting the category.
    n1Group = records_by_item[row[0]]['category']
    n2Group = records_by_item[row[1]]['category']

    item = { 'n1': row[0], 'n2': row[1], 'score': row[2],  'c1': n1Group, 'c2': n2Group }

    # If we don't have an entry for n1, the current phenotype profile being 
    #  considered, create an try for it in the dictionary of top scores
    if not item['n1'] in topScores:
      topScores[item['n1']] = item
    
    # If the score of the current comparison is greater than the existing top score recorded 
    #   for this phenotype profile, replace it with this one
    if item['score'] > topScores[item['n1']]['score']:
      topScores[item['n1']] = item

print('Done!')

In [ ]:
# Here we're creating two lists, both with one entry for each phenotype profile:
#  truth: which group does the phenotype profile belong to?
#  pred: which group is this phenotype profile predicted to belong to (i.e. the group of its most similar phenotype profile)?
truth = list()
pred = list()

for k in topScores:
  s = topScores[k]
  truth.append(s['c1'])
  pred.append(s['c2'])
    
# Now we can calculate precision
precision = sklearn.metrics.precision_score(truth, pred, average='weighted', zero_division=1)

print('Precision: ' + str(precision))

Here we get a much lower precision. But that's okay, really. We have to consider that this is a non-binary problem, so being correct around 50% of the time is still much better than random. Indeed, one of the 'problems' of semantic similarity is that it's essentially undirected and unsupervised. We're measuring the similarity of these entities, but we don't necessarily know the basis by which those scores are derived. It's likely that, while certain phenotypes are more associated with certain note categories, phenotypic similarity is more clearly aligned to patients. 

> Bonus tasks: 
> * What else can we predict using semantic similarity?
> * What other ways could we transform the pairwise similarity matrix into a set of predictions?

# Further Reading

* [Komenti Application Note](https://www.biorxiv.org/content/10.1101/2020.08.04.233049v1)
* [Klarigi Application Note](https://www.biorxiv.org/content/10.1101/2021.06.14.448423v1)
* [Towards Similarity-based Differential Diagnostics for Common Diseases](https://www.sciencedirect.com/science/article/pii/S0010482521001542)
* [Multi-faceted Semantic Clustering With Text-derived Phenotypes](https://www.medrxiv.org/content/10.1101/2021.05.26.21257830v1)
* [Other Komenti Guides](https://github.com/reality/komenti_guide)
* [Komenti on GitHub](https://github.com/reality/Komenti)
* [Klarigi on GitHub](https://github.com/reality/Klarigi)
